packages/hebao_v1/contracts/modules/dapps/CompoundModule.sol

```diff
@@ -85,13 +118,62 @@ contract CompoundModule is BaseSubAccount, SecurityModule
         address cToken = compoundRegistry.getCToken(token);
         require(cToken != address(0), "NO_MARKET");
 
-        uint balance = uint(tokenBalance(wallet, token));
-        require(amount > 0 && amount <= balance, "INVALID_WITHDRAW_AMOUNT");
+        require(amount > 0, "ZERO_WITHDRAW_AMOUNT");
+
+        if (isCollaterlCToken(wallet, cToken)) {
+            // for collaterl token, check liquidity to withdraw.
+            (uint collateralRemovable, uint borrowableAfterRemove) = collateralTokenRemovable(wallet, token);
+            require(amount <= collateralRemovable + borrowableAfterRemove, "INVALID_WITHDRAW_AMOUNT");
+
+            uint redeemable = collateralRemovable;
+            uint borrowable;
+            if (amount <= collateralRemovable) {
+                redeemable = amount;
+            } else {
+                borrowable = amount - collateralRemovable;
+            }
+
+            if (redeemable > 0) {
+                redeemUnderlying(wallet, cToken, redeemable);
+                trackWithdrawal(wallet, token, redeemable);
+            }
 
-        redeem(wallet, cToken, amount);
-        trackWithdrawal(wallet, token, amount);
+            if (borrowable > 0) {
+                borrow(wallet, cToken, borrowable);
+                trackBorrow(wallet, token, borrowable);
+            }
+            return;
```

Comment:
It's a bit confusing there's a `return` here. Can you make it an if/else instead?

Reply:
fixed